doc/policy_based_metering/PBM-HLD.md

```diff
@@ -0,0 +1,352 @@
+
+Usage examples:
+- **Security**: Policy-Based Metering can be used to guard against network storms or DDoS attacks by limiting traffic volumes.
+- **Quality of Service (QoS)**: It helps in prioritizing critical traffic (like VoIP or video streams) while limiting less important traffic (such as bulk data transfers).
+- **Data Centre**: ensure fair bandwidth distribution.
+
+##### *Image 1: Traffic Flow Overview*
+![alt text](Overview-Flowchart.jpg)
+
+---
+### Requirements
+#### Functional Requirements
+- Backward compatibility for existing ACL features - If policer is not set, the system will function as it did previously.
+- Ability to config/remove/manage policers with ACL entries.
+- Support policer types
+	- **storm_control**: Used to prevent network storms by limiting the rate of broadcast, multicast, and unknown unicast traffic.
+	- **srTCM (Single Rate Three Color Marker)**: Marks packets based on a single rate and three colors (green, yellow, red) to indicate compliance with the traffic profile.
+	- **trTCM (Two Rate Three Color Marker)**: Marks packets based on two rates (committed and peak) and three colors to indicate compliance.
+	- Support rate-limiting in terms of packets per second or bytes per second.
+- Support all existing ACL types (ACL table types, ACL stages).
+### Scalability Requirements:
+- Support for multiple ACL entries with associated policers.
+- Query and validate ASIC capabilities dynamically.
+#### CLI Requirements
+- Bind policers with ACL rules.
+- Unbind policer from ACL rules.
+- Show ACLs with policers
+---
+### Architecture Design
+
+No SONiC architecture changes are required as the existing infrastructure is being used.
+##### *Image 2: ACL Configuration Flow Overview*
+
+![alt text](Overview.jpg)
+
+---
+### High-Level Design
+
+#### Modules and Sub-Modules
+- **SWSS**
+	- ACL-Orch
+		- Add/Remove/Mange ACLs.
+		- Set or disable policer action.
+		- Query from SAI the ACL actions capability.
+		- Allow policer action only when capability is enabled (SAI query).
+		- As part of rule processing
+		    - Parse and validate policer info as part of rule action processing.
+	- Policer-Orch
+		- Add/Remove/Mange Policers.
+
+##### *Image 3: Configuration Flow Overview*
+
+![[Config_Flow.jpg]]
+
+---
+### SAI API
+| SAI Attribute                           | Description                      |
+| --------------------------------------- | -------------------------------- |
+| set_acl_entry_attribute()               | Attach policers to ACL rules     |
+| sai_query_attribute_capability()        | Verify ASIC support for policers |
+| SAI_ACL_ACTION_TYPE_SET_POLICER         |                                  |
+| SAI_ACL_ACTION_TYPE_PACKET_ACTION       |                                  |
+| SAI_ACL_ENTRY_ATTR_ACTION_PACKET_ACTION |                                  |
+| SAI_ACL_ENTRY_ATTR_ACTION_SET_POLICER   |                                  |
+
```

Comment:
these are APIs that already exist? are they already being used? whatever the case, this should be stated here. (in a sentence - adding support for these APIs...)